mozci/data/sources/treeherder/__init__.py

```diff
@@ -18,15 +21,31 @@
     Push = None
 
 
-class BaseTreeherderSource(DataSource):
-    pass
+class BaseTreeherderSource(DataSource, ABC):
+    lock = threading.Lock()
+    groups_cache: Dict[str, List[str]] = {}
```

Comment:
The only drawback I see here is that this will leak memory. We might want to delete the tasks from the cache once we get the groups from them (IIRC we only get them once anyway)

Reply:
I added an LRU dict library to cache this (I added a dependency rather than roll our own since it's implemented in C and is faster)